### PR TITLE
Hmis loader performance

### PR DESCRIPTION
**Pull request by amichal**
- [x] Switch the initial load! from `activerecord-import` to PG `COPY DATA`  while still  preprocessing lines/headers in ruby to guess encoding and inject meta data  (~5x improvement on my benchmark)
- [x] Add detailed timing information to the debug logs
- [x] Remove preprocessed line counts, since the CSV can contain embedded newlines they don't tell us much
- [x] Cut down on allocations, precompute what we can and avoid newing up AR models where we can on import! (~3x improvement on my benchmark)
- [x] Tamper with `fix_time_format` used by importer to return Time/DateTime objects so import! does not screw up the default timezone

### CODE REVIEW

**Comment by amichal** [comment]
Testing  this weekend with more real world data found a few things I'm fixing.

File: drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb [hunk 1]
```diff
@@ -205,20 +213,22 @@ def self.fix_date_format(string)
...
-      return string.to_time.strftime('%Y-%m-%d %H:%M:%S') if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
+      return Time.zone.parse(string) if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
 
       # Sometimes times come in mm-dd-yyyy hh:mm
       if /\d{1,2}-\d{1,2}-\d{4} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
         date, time = string.split(' ')
         month, day, year = date.split('-')
-
-        return "#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}".to_time.strftime('%Y-%m-%d %H:%M:%S')
+        return Time.zone.parse("#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}")
```
Comment (eanders) [inline drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb]: I think this isn't picking the right timezone.  On the left, previous behavior, on the right, current behavior (loaded on top, imported on bottom)
Screen Shot 2021-02-03 at 9 22 31 AM

**Comment by amichal** [comment]
Ugh... Maybe a testing, `travel_to` interaction with AR::Import? Not sure
why else it passed tests locally which look for this.

**Comment by amichal** [comment]
I'll look into it.

File: drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb [hunk 2]
```diff
@@ -167,102 +167,83 @@ def self.accepted_time_pattern
...
+    ].freeze
+
+    TIME_FORMATS = ([
+      ['%Y-%m-%d %H:%M:%S', RE_YYYYMMDD],
+      ['%m-%d-%Y %H:%M:%S'],
+      ['%d-%b-%Y %H:%M:%S'],
+      ['%Y-%m-%d %H:%M', RE_YYYYMMDD],
+      ['%m-%d-%Y %H:%M'],
+      ['%d-%b-%Y %H:%M'],
+    ] + DATE_FORMATS).freeze # order matters, we need to try more logical and longer patterns first
```
Comment (eanders) [inline drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb]: This is a great idea.

File: drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/loader/loader.rb [hunk 3]
```diff
@@ -300,19 +327,23 @@ def export_file_valid?
...
-    end
-
-    def loaded_lines(file, count)
-      @loader_log.summary[file]['lines_loaded'] += count
+      @loader_log.update(
+        completed_at: Time.current,
+        status: status,
+      )
+      status = "#{status} error:#{err}" if err
+      log("Completed in #{elapsed_time(elapsed)} #{hash_as_log_str log_ids}. status:#{status}\n#{summary_as_log_str loader_log.summary}")
```
Comment (eanders) [inline drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/loader/loader.rb]: Let's make this `Completed loading in ....`

File: drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/importer.rb [hunk 4]
```diff
@@ -592,26 +678,39 @@ def self.soft_deletable_sources
...
+      end
+    end
+
+    def complete_import
+      data_source.update(last_imported_at: Time.zone.now)
+      importer_log.completed_at = Time.zone.now
+      importer_log.upload_id = @upload.id if @upload.present?
+      importer_log.save
+      elapsed = Time.current - @started_at
+      log("Completed in #{elapsed_time(elapsed)} #{hash_as_log_str log_ids}.\n#{summary_as_log_str importer_log.summary}")
```
Comment (eanders) [inline drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/importer.rb]: Let's make this `Completed importing in ....`

**Pull request merged**